pkg/deploy/controller/deployment/factory.go

```diff
@@ -1,177 +1,205 @@
-		}
-		for _, env := range strategy.CustomParams.Environment {
-			set.Insert(env.Name)
-			environment = append(environment, env)
+// Run begins watching and syncing.
+func (c *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Wait for the dc store to sync before starting any work in this controller.
+	ready := make(chan struct{})
+	go c.waitForSyncedStores(ready, stopCh)
+	select {
+	case <-ready:
+	case <-stopCh:
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down deployment controller")
+	c.queue.ShutDown()
+}
+
+func (c *DeploymentController) waitForSyncedStores(ready chan<- struct{}, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	for !c.rcStoreSynced() || !c.podStoreSynced() {
+		glog.V(4).Infof("Waiting for the rc and pod caches to sync before starting the deployment controller workers")
+		select {
+		case <-time.After(StoreSyncedPollPeriod):
+		case <-stopCh:
+			return
 		}
 	}
+	close(ready)
+}
+
+func (c *DeploymentController) addReplicationController(obj interface{}) {
+	rc := obj.(*kapi.ReplicationController)
+	// Filter out all unrelated replication controllers.
+	if !deployutil.IsOwnedByConfig(rc) {
+		return
+	}
+
+	c.enqueueReplicationController(rc)
+}
+
+func (c *DeploymentController) updateReplicationController(old, cur interface{}) {
+	rc := cur.(*kapi.ReplicationController)
+	// Filter out all unrelated replication controllers.
+	if !deployutil.IsOwnedByConfig(rc) {
+		return
+	}
+
+	c.enqueueReplicationController(rc)
+}
+
+func (c *DeploymentController) updatePod(old, cur interface{}) {
+	// A periodic relist will send update events for all known controllers.
+	if kapi.Semantic.DeepEqual(old, cur) {
+		return
+	}
+
+	pod := cur.(*kapi.Pod)
+	// Filter out all but deployer pods
+	if !deployutil.IsOwnedByDeployment(pod) {
+		return
+	}
+
+	if rc, err := c.rcForDeployerPod(pod); err == nil && rc != nil {
+		c.enqueueReplicationController(rc)
+	}
+}
+
+func (c *DeploymentController) deletePod(obj interface{}) {
+	pod := obj.(*kapi.Pod)
+	// Filter out all but deployer pods
+	if !deployutil.IsOwnedByDeployment(pod) {
+		return
+	}
+
+	if rc, err := c.rcForDeployerPod(pod); err == nil && rc != nil {
+		c.enqueueReplicationController(rc)
+	}
+}
+
+func (c *DeploymentController) enqueueReplicationController(rc *kapi.ReplicationController) {
+	key, err := kcontroller.KeyFunc(rc)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %#v: %v", rc, err)
+		return
+	}
+	c.queue.Add(key)
+}
 
-	// Set default environment values
-	for _, env := range factory.Environment {
-		if set.Has(env.Name) {
-			continue
+func (c *DeploymentController) rcForDeployerPod(pod *kapi.Pod) (*kapi.ReplicationController, error) {
+	key := pod.Namespace + "/" + deployutil.DeploymentNameFor(pod)
+	return c.getByKey(key)
+}
+
+func (c *DeploymentController) worker() {
+	for {
```

Comment:
But it's not good both health- and mental-wise

:)

On Mon, Jun 27, 2016 at 2:09 PM, Michal Fojtik <notifications@github.com>
wrote:

> In pkg/deploy/controller/deployment/factory.go
> <https://github.com/openshift/origin/pull/9327#discussion_r68565227>:
>
> >
> > -	// Set default environment values
> > -	for _, env := range factory.Environment {
> > -		if set.Has(env.Name) {
> > -			continue
> > +func (c *DeploymentController) rcForDeployerPod(pod *kapi.Pod) (*kapi.ReplicationController, error) {
> > +	key := pod.Namespace + "/" + deployutil.DeploymentNameFor(pod)
> > +	return c.getByKey(key)
> > +}
> > +
> > +func (c *DeploymentController) worker() {
> > +	for {
>
> @kargakis <https://github.com/kargakis> I like that workers stay in work
> even the assembly line shut downs... reminds me us ;-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9327/files/2bc801bef0539b5191b4d473cdf2a34f2ce817cd#r68565227>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADuFf01MFStzCBcVkYwqfzTWjfvO9dUnks5qP713gaJpZM4I1iD4>
> .
>